### ISSUE TITLE
 UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead

### ISSUE THREAD

**Issue by anusurendra** [deepchem/deepchem#1464]
**Description:**
- DeepChem Version: 2.1.0
- Platform: Ubuntu 16.04

I am following the Predicting Ki of Ligands to a Protein tutorial (https://deepchem.io/docs/notebooks/BACE.html), on my data. However, I am getting following error: ` warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.`

I am not sure if this a known bug? I have iterated through my raw dataset and found no NaN that could lead to this error.

Below is the code:

```
filepath_for_data = sys.argv[1]
file_index_for_data = sys.argv[2]
current_dir = os.path.dirname(os.path.realpath("__file__"))
#dc.utils.download_url("https://s3-us-west-1.amazonaws.com/deepchem.io/datasets/desc_canvas_aug30.csv",
#                      current_dir)
dataset_file = filepath_for_data+"/trainset/"+file_index_for_data+"_train.csv"
dataset = load_from_disk(dataset_file)
num_display=10
pretty_columns = (
    "[" + ",".join(["'%s'" % column for column in dataset.columns.values[:num_display]])
    + ",...]")

#dc.utils.download_url("https://s3-us-west-1.amazonaws.com/deepchem.io/datasets/crystal_desc_canvas_aug30.csv",
 #                     current_dir)
crystal_dataset_file = filepath_for_data+"/predictset/"+file_index_for_data+"_predict.csv"
crystal_dataset = load_from_disk(crystal_dataset_file)

print("Columns of dataset: %s" % pretty_columns)
print("Number of examples in dataset: %s" % str(dataset.shape[0]))
print("Number of examples in crystal dataset: %s" % str(crystal_dataset.shape[0]))


user_specified_features = ['MW','AlogP','HBA','HBD','RB','HeavyAtomCount',...,'PEOE13 (PEOE13)','PEOE14 (PEOE14)']


featurizer = dc.feat.UserDefinedFeaturizer(user_specified_features)
loader = dc.data.UserCSVLoader(
      tasks=["Class"], smiles_field="mol", id_field="mol",
      featurizer=featurizer)
dataset = loader.featurize(dataset_file)
crystal_dataset = loader.featurize(crystal_dataset_file)


splitter = dc.splits.SpecifiedSplitter(dataset_file, "Model")
train_dataset, valid_dataset, test_dataset = splitter.train_valid_test_split(
    dataset)
#NOTE THE RENAMING:
valid_dataset, test_dataset = test_dataset, valid_dataset


for x, y, w, id in valid_dataset.itersamples():
    print(x, y, w, id)


transformers = [
    dc.trans.NormalizationTransformer(transform_X=True, dataset=train_dataset),
    dc.trans.ClippingTransformer(transform_X=True, dataset=train_dataset)]

datasets = [train_dataset, valid_dataset, test_dataset, crystal_dataset]
for i, dataset in enumerate(datasets):
  for transformer in transformers:
      datasets[i] = transformer.transform(dataset)
train_dataset, valid_dataset, test_dataset, crystal_dataset = datasets


def rf_model_builder(model_params, model_dir):
  sklearn_model = RandomForestClassifier(**model_params)
  return dc.models.SklearnModel(sklearn_model, model_dir)



params_dict = {
    "n_estimators": [10, 100],
    "max_features": ["auto", "sqrt", "log2", None],
}

metric = dc.metrics.Metric(dc.metrics.roc_auc_score)
optimizer = dc.hyper.HyperparamOpt(rf_model_builder)
best_rf, best_rf_hyperparams, all_rf_results = optimizer.hyperparam_search(
    params_dict, train_dataset, valid_dataset, transformers,
    metric=metric)


params_dict = {"learning_rate": np.power(10., np.random.uniform(-5, -3, size=1)),
               "weight_decay_penalty": np.power(10, np.random.uniform(-6, -4, size=1)),
               "nb_epoch": [40] }
n_features = train_dataset.get_data_shape()[0]
def model_builder(model_params, model_dir):
  model = dc.models.MultiTaskClassifier(
    1, n_features, layer_sizes=[1000], dropouts=.25,
    batch_size=50, **model_params)
  return model

optimizer = dc.hyper.HyperparamOpt(model_builder)
best_dnn, best_dnn_hyperparams, all_dnn_results = optimizer.hyperparam_search(
    params_dict, train_dataset, valid_dataset, transformers,
    metric=metric)



rf_train_csv_out = "rf_train_regressor.csv"
rf_train_stats_out = "rf_train_stats_regressor.txt"
rf_train_evaluator = Evaluator(best_rf, train_dataset, transformers)
rf_train_score = rf_train_evaluator.compute_model_performance(
    [metric], rf_train_csv_out, rf_train_stats_out)
print("RF Train set AUC %f" % (rf_train_score["roc_auc_score"]))

rf_valid_csv_out = "rf_valid_regressor.csv"
rf_valid_stats_out = "rf_valid_stats_regressor.txt"
rf_valid_evaluator = Evaluator(best_rf, valid_dataset, transformers)
rf_valid_score = rf_valid_evaluator.compute_model_performance(
    [metric], rf_valid_csv_out, rf_valid_stats_out)
print("RF Valid set AUC %f" % (rf_valid_score["roc_auc_score"]))

rf_test_csv_out = "rf_test_regressor.csv"
rf_test_stats_out = "rf_test_stats_regressor.txt"
rf_test_evaluator = Evaluator(best_rf, test_dataset, transformers)
rf_test_score = rf_test_evaluator.compute_model_performance(
    [metric], rf_test_csv_out, rf_test_stats_out)
print("RF Test set AUC %f" % (rf_test_score["roc_auc_score"]))

rf_crystal_csv_out = "rf_crystal_regressor.csv"
rf_crystal_stats_out = "rf_crystal_stats_regressor.txt"
rf_crystal_evaluator = Evaluator(best_rf, crystal_dataset, transformers)
rf_crystal_score = rf_crystal_evaluator.compute_model_performance(
    [metric], rf_crystal_csv_out, rf_crystal_stats_out)
print("RF Crystal set R^2 %f" % (rf_crystal_score["roc_auc_score"]))

```

Below is the full error:

```
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/trans/transformers.py:148: RuntimeWarning: invalid value encountered in true_divide
  X = np.nan_to_num((X - self.X_means) / self.X_stds)
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/trans/transformers.py:148: RuntimeWarning: divide by zero encountered in true_divide
  X = np.nan_to_num((X - self.X_means) / self.X_stds)
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
/home/surendraa/software/anaconda3/envs/snowflakes040518/lib/python3.5/site-packages/deepchem/metrics/__init__.py:312: UserWarning: Error calculating metric roc_auc_score: Expected 2D array, got 1D array instead:
array=[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
  warnings.warn("Error calculating metric %s: %s" % (self.name, e))
Columns of dataset: ['mol','CID','Class','Model','pIC50','MW','AlogP','HBA','HBD','RB',...]
Number of examples in dataset: 263
Number of examples in crystal dataset: 1
Loading raw samples now.
shard_size: 8192
About to start loading CSV from /home/surendraa/Documents/10027278_train.csv
Loading shard 1 of size 8192.
TIMING: user specified processing took 0.107 s
TIMING: featurizing shard 0 took 0.107 s
TIMING: dataset construction took 0.156 s
Loading dataset from disk.
Loading raw samples now.
shard_size: 8192
About to start loading CSV from /home/surendraa/Documents/10027278_predict.csv
Loading shard 1 of size 8192.
TIMING: user specified processing took 0.106 s
TIMING: featurizing shard 0 took 0.106 s
TIMING: dataset construction took 0.134 s
Loading dataset from disk.
TIMING: dataset construction took 0.018 s
Loading dataset from disk.
TIMING: dataset construction took 0.007 s
Loading dataset from disk.
TIMING: dataset construction took 0.009 s
Loading dataset from disk.
TIMING: dataset construction took 0.021 s
Loading dataset from disk.
TIMING: dataset construction took 0.014 s
Loading dataset from disk.
TIMING: dataset construction took 0.007 s
Loading dataset from disk.
TIMING: dataset construction took 0.006 s
Loading dataset from disk.
TIMING: dataset construction took 0.005 s
Loading dataset from disk.
TIMING: dataset construction took 0.004 s
Loading dataset from disk.
TIMING: dataset construction took 0.004 s
Loading dataset from disk.
TIMING: dataset construction took 0.004 s
Loading dataset from disk.
Fitting model 1/8
hyperparameters: {'n_estimators': 10, 'max_features': 'auto'}
computed_metrics: [nan]
Model 1/8, Metric roc_auc_score, Validation set 0: nan
	best_validation_score so far: -inf
Fitting model 2/8
hyperparameters: {'n_estimators': 10, 'max_features': 'sqrt'}
computed_metrics: [nan]
Model 2/8, Metric roc_auc_score, Validation set 1: nan
	best_validation_score so far: -inf
Fitting model 3/8
hyperparameters: {'n_estimators': 10, 'max_features': 'log2'}
computed_metrics: [nan]
Model 3/8, Metric roc_auc_score, Validation set 2: nan
	best_validation_score so far: -inf
Fitting model 4/8
hyperparameters: {'n_estimators': 10, 'max_features': None}
computed_metrics: [nan]
Model 4/8, Metric roc_auc_score, Validation set 3: nan
	best_validation_score so far: -inf
Fitting model 5/8
hyperparameters: {'n_estimators': 100, 'max_features': 'auto'}
computed_metrics: [nan]
Model 5/8, Metric roc_auc_score, Validation set 4: nan
	best_validation_score so far: -inf
Fitting model 6/8
hyperparameters: {'n_estimators': 100, 'max_features': 'sqrt'}
computed_metrics: [nan]
Model 6/8, Metric roc_auc_score, Validation set 5: nan
	best_validation_score so far: -inf
Fitting model 7/8
hyperparameters: {'n_estimators': 100, 'max_features': 'log2'}
computed_metrics: [nan]
Model 7/8, Metric roc_auc_score, Validation set 6: nan
	best_validation_score so far: -inf
Fitting model 8/8
hyperparameters: {'n_estimators': 100, 'max_features': None}
computed_metrics: [nan]
Model 8/8, Metric roc_auc_score, Validation set 7: nan
	best_validation_score so far: -inf
No models trained correctly.
```

Any help would be much appreciated.

**Comment by anusurendra** [comment]
It seems that after the transforming the data to normalize it to have zero-mean and unit-standard-deviation. When I iterate through the train_set, the x is an array of length 589 instead of an array with a length of 1. I am not sure why this would be the case?

```
transformers = [
    dc.trans.NormalizationTransformer(transform_X=True, dataset=train_dataset),
    dc.trans.ClippingTransformer(transform_X=True, dataset=train_dataset)]

datasets = [train_dataset, valid_dataset, test_dataset, crystal_dataset]
for i, dataset in enumerate(datasets):
  for transformer in transformers:
      datasets[i] = transformer.transform(dataset)
train_dataset, valid_dataset, test_dataset, crystal_dataset = datasets

for x, y, w, id in train_dataset.itersamples():
    print(x, y, w, id)

```

**Comment by VIGS25** [comment]
Hi, 
I would presume X is ordered as (`n_samples`, `n_features`) so each sample should have a shape of `n_features`. So the 589 is possible right?

**Comment by chf42** [comment]
@anusurendra Hi, I came across the same issue. The error happened when I tried to use transform for my testset. /deepchem/trans/transformers.py:148: RuntimeWarning: invalid value encountered in true_divide .  Have you figured a way out? Thanks

**Comment by anusurendra** [comment]
@chf42, sorry for the delayed response. Unfortunately we did not figure out the issue. But this is motivation for me to get back to the issue. I will let you if I get anywhere.
